### ISSUE TITLE
Wrong group assigned, lag with lots of players?

### ISSUE THREAD

**Issue by LemADEC** [PEXPlugins/PermissionsEx#1881]
When resetting a player on a ladder, we're quickly sending 2 commands: one to remove players from add a new group and another to remove its current group on that ladder.
This worked fine so far with PermissionsEx-1.22.6 on MC 1.7.10.
Since we've switched toPermissionsEx-1.23.1 on MC 1.8.1 it seems to get mixed up and player ends up with no group assigned or both groups assigned. It feels like the second permission starts processing while the first one isn't finished, ending up with invalid perms saved.
This issue is reproduced with below configuration and 7000+ users added to it.
Issue is not reproduced when only 1 user is present.

Here's the configuration without the extra 7k users:
```
groups:
  Member:
    permissions:
    - nothing
  AAA:
    inheritance:
    - Member
    permissions:
    - nothing
  BBB:
    inheritance:
    - AAA
    permissions:
    - nothing
users:
```

Here are the commands used to reproduce it, just past them on console:
```
pex user lemadec group add AAA
pex user lemadec group remove BBB
pex user lemadec
pex user lemadec group remove AAA
pex user lemadec group add BBB
pex user lemadec
```

and the resulting logs (UUIDs redacted for sake of readability):
```
>pex user lemadec group add AAA
>pex user lemadec group remove BBB
>pex user lemadec
>pex user lemadec group remove AAA
>pex user lemadec group add BBB
>pex user lemadec
[12:25:38 INFO]: User "xxx/LemADEC" added to group "AAA"!
[12:25:38 INFO]: User "xxx/LemADEC" removed from group "BBB"!
[12:25:38 INFO]: 'xxx/LemADEC' is a member of:
[12:25:38 INFO]:    AAA (not ranked)
[12:25:38 INFO]:    BBB (not ranked)
[12:25:38 INFO]: LemADEC's permissions:
[12:25:38 INFO]:
[12:25:38 INFO]: LemADEC's options:
[12:25:38 INFO]:   name = "LemADEC"
[12:25:39 INFO]: User "xxx/LemADEC" removed from group "AAA"!
[12:25:39 INFO]: User "xxx/LemADEC" added to group "BBB"!
[12:25:39 INFO]: 'xxx/LemADEC' is a member of:
[12:25:39 INFO]:    AAA (not ranked)
[12:25:39 INFO]:    BBB (not ranked)
[12:25:39 INFO]: LemADEC's permissions:
[12:25:39 INFO]:
[12:25:39 INFO]: LemADEC's options:
[12:25:39 INFO]:   name = "LemADEC"
```

**Comment by CullanP** [comment]
Yes! Something is seriously wrong with either pex or vault. I'm hooked into vault and using "for (String pg : RankupPlugin.perms.getPlayerGroups(player)) {" This returns the wrong groups and normally keeps returning my default group. I'm not sure why this is happening, but it only recently started to happen when I updated both vault and pex. Please look into this @zml2008

**Comment by zml2008** [comment]
Do you only experience this with users who are offline? And what is the output if you enable debug mode during some of these changes? I think there might be architectural issues involved here, so you might just have to wait for pex2.0 to be ready

**Comment by LemADEC** [comment]
It was only tested with online players.
For the record, Pex fails to get proper UUID on offline players, so it's not an option.
What's the ETA for pex2.0?

**Comment by CullanP** [comment]
I fixed the issue zml, I know 3 files that the issue could lie in(CachingData, CachingGroupData, or CachingUserData). The reason I don't know which class exactly the issue lies in, was due to me grabbing pex 1.2.6, decompiling it and grabbing those 3 files and putting them in. This solved my issue. The ranks are not updating and grabbing groups as intended. If you want to figure out what's wrong I'd look there. :)
